mlfinlab/online_portfolio_selection/online_portfolio_selection.py

```diff
@@ -0,0 +1,327 @@
+        # Calculate returns by taking the cumulative product.
+        self.portfolio_return = np.diagonal(np.dot(relative_return, all_weights.T)).cumprod()
+
+    def conversion(self):
+        """
+        Converts the given np.array to pd.Dataframe.
+        """
+        # Convert all_weights.
+        self.all_weights = pd.DataFrame(self.all_weights, index=self.time, columns=self.asset_name)
+
+        # Convert portfolio_return.
+        self.portfolio_return = pd.DataFrame(self.portfolio_return, index=self.time, columns=["Returns"])
+
+    def optimize(self, optimize_array, solver=cp.SCS):
+        """
+        Calculates weights that maximize returns over the given array.
+
+        :param optimize_array: (np.array) Relative returns of the assets for a given time period.
+        :param solver: (cp.solver) Solver for cvxpy
+        :return weights.value: (np.array) Weights that maximize the returns for the given array.
+        """
+
+        # Initialize weights for the optimization problem.
+        weights = cp.Variable(self.number_of_assets)
+
+        # Use cp.log and cp.sum to make the cost function a convex function.
+        # Multiplying continuous returns equates to summing over the log returns.
+        portfolio_return = cp.sum(cp.log(optimize_array * weights))
+
+        # Optimization objective and constraints.
+        allocation_objective = cp.Maximize(portfolio_return)
+        allocation_constraints = [cp.sum(weights) == 1, cp.min(weights) >= 0]
+
+        # Define and solve the problem.
+        problem = cp.Problem(objective=allocation_objective, constraints=allocation_constraints)
+
+        # Solve and return the resulting weights.
+        problem.solve(warm_start=True, solver=solver)
+        return weights.value
+
+    def round_weights(self, threshold=1e-6):
+        """
+        Drops weights that are below a certain threshold.
+
+        :param threshold: (float) Drop all values below this threshold.
+        """
+        # Set all values below the threshold to 0.
+        new_all_weights = np.where(self.all_weights < threshold, 0, self.all_weights)
+
+        # Adjust weights to have the weights sum to 1.
+        new_all_weights = np.apply_along_axis(lambda x: x / np.sum(x), 1, new_all_weights)
+        self.all_weights = new_all_weights
+
+    def uniform_weight(self):
+        """
+        Returns a uniform weight of assets.
+
+        :return uni_weight: (np.array) Uniform weights (1/n, 1/n, 1/n ...).
+        """
+        # Divide by number of assets after creating numpy arrays of one.
+        uni_weight = np.ones(self.number_of_assets) / self.number_of_assets
+        return uni_weight
+
+    def print_progress(self, iteration, prefix='', suffix='', decimals=1, bar_length=50):
+        # pylint: disable=expression-not-assigned
+        """
+        Calls in a loop to create a terminal progress bar.
+        https://gist.github.com/aubricus/f91fb55dc6ba5557fbab06119420dd6a
```

Comment:
What is this link? `https://gist.github.com/aubricus/f91fb55dc6ba5557fbab06119420dd6a`

Reply:
it's a progress bar method that I found online